context: Fix hotplug handling and signal emission during enumerate

benzea opened the pull request:
Initial device enumeration may happen in a random order. This causes
issues if someone calls g_usb_device_get_parent from the "device-added"
signal handler. The problem is. that this will cause an infinite
recursion as GUsb tries to enumerate again, which results in the device
to be added. Also, the order of devices is random and resolving the
parent requires the full list of devices. As such, we should delay
emission until the internal array has been filled.

To fix this, emit the signals from g_usb_context_enumerate.

Also, while changing this, it became obvious that libusb hotplug callback
is updating priv->devices from the libusb thread, only pushing it to an
idle handler when the signal is emitted. Doing this is unsafe if hotplug
events arrive during enumeration. Push libusb event into the mainloop
immediately in order to process it.

Doing means that any libusb hotplug events arriving during
g_usb_context_enumerate will be handled after the function has returned.
Making all possible corner cases behave correctly.

hughsie commented on gusb/gusb-context.c:
```diff
@@ -242,59 +242,26 @@ g_usb_context_class_init (GUsbContextClass *klass)
-	helper->signal_id = DEVICE_ADDED_SIGNAL;
-	if (!context->priv->done_enumerate) {
-		g_usb_context_idle_signal_cb (helper);
+	/* Emitted directly by g_usb_context_enumerate */
```
lower case Emitted please

hughsie commented on gusb/gusb-context.c:
```diff
@@ -242,59 +242,26 @@ g_usb_context_class_init (GUsbContextClass *klass)
-	helper->signal_id = DEVICE_REMOVED_SIGNAL;
-	if (!context->priv->done_enumerate) {
-		g_usb_context_idle_signal_cb (helper);
+	/* Cannot happen, if it did we would not fire any signal. */
```
lower case comment, and no trailing fullstop please

hughsie commented on gusb/gusb-context.c:
```diff
@@ -538,6 +538,13 @@ g_usb_context_enumerate (GUsbContext *context)
 	priv->done_enumerate = TRUE;
+
+	/* Emit device-added signals before returning. */
+	for (i = 0; i < priv->devices->len; i++)
```
can you put the `i` in the for loop pls.

hughsie commented on gusb/gusb-context.c:
```diff
@@ -538,6 +538,13 @@ g_usb_context_enumerate (GUsbContext *context)
+		g_signal_emit (context, signals[DEVICE_ADDED_SIGNAL], 0,
+			       g_ptr_array_index (priv->devices, i));
+
+	/* Any queued up hotplug events are queued as idle handlers. */
```
same with the comments pls, then LGTM.

benzea commented:
OK, updated.

I changed one of the comments slightly. I think there is a theoretical change in behaviour. i.e. before libusb hotplug events could come through to the application even before `g_usb_context_enumerate` was called. Now, hotplug events can only happen during/after `g_usb_context_enumerate`.

I think that is fine, just wanted to point it out in case it might be relevant.

hughsie commented:
> I think that is fine, just wanted to point it out in case it might be relevant.

I think you fixed a bug! Thanks!
